---
layout: post
title:  "Differentially expressed gene (DSC2) in cell clusters by k-means"
author: April Yan
jhed: yyan67
categories: [ HW6 ]
image: homework/hw6/hw6_yyan67.png
featured: false
---

### Perform a full analysis (quality control, dimensionality reduction, kmeans clustering, differential expression analysis) on your data. Your goal is to figure out what tissue structure is represented in the CODEX data. Options include: (1) Artery/Vein, (2) White pulp, (3) Red pulp, (4) Capsule/Trabecula You will need to visualize and interpret at least two cell-types. Create a data visualization and write a description to convince me that your interpretation is correct. 

### Your description should reference papers and content that allowed you to interpret your cell clusters as a particular cell-types. You must provide attribution to external resources referenced. Links are fine; formatted references are not required. 

I performed k-means clustering on t-SNE embedding of normalized gene expression data. The very top panel showed an elbow plot to find an optimal k. I found k=6 could be optimal since the total witness stopped dropping drastically after k=6. We can also see how cells are distributed in t-SNE with it k-means labels (top right panel). I picked cluster 1 and 6 to further explore their cell types. I did Wilcoxon tests for differential expression analysis in cluster 1 and 6 respectively. The second row corresponds to cluster 1 and the third row corresponds to cluster 6. I plotted their log10 of pvalues vs. log2 fold change to identify significantly expressed genes. 

"Lyve1" seem to be one of those in cluster 1, which are shown to be significantly expressed in Lymphatic endothelial cells (LECs) that form the structure of the **lymphatic vessels** [1,2]. I have plotted cluster 1 cells in physical space. If this is an image of a slice of spleen tissue, it will make sense that we observe the circle-shaped structures, which is just the cross section of vessels/veins. 

"CD3e" and "CD45" seem to be significant highly expressed genes in cluster 6. Gene names starts with "CD" are usually related to immune cells. I think this cluster belongs to red pulp cells due to the following reasons. First, CD45 regulates red pulp macrophages and erythropoietin in the spleen [3]. Second, this can't be white pulp because CD3e can't be detected in white pulp cells [4]. CD45RO and CD4 seem to be one of those in cluster 6. I have labeled the gene expression level of CD45 in third row third panel and we can see it's highly expressed in cluster 6 (red pulp). 

## Please share the code you used to reproduce this data visualization.
```{r}
## HW6
# rm(list=ls())
data <- read.csv('codex_spleen_subset.csv.gz', row.names = 1)

library(ggplot2)
library(patchwork)
library(Rtsne)

### quality control
# -- low variance filtering and normalization
set.seed(1111)
gexp <- data[, 4:ncol(data)]
gexpnorm <- log10(gexp/rowSums(gexp) * mean(rowSums(gexp))+1)

### dimension reduction 
# -- t-SNE embedding on PCs
# pcs <- prcomp(gexpnorm)
# emb <- Rtsne(pcs$x)$Y

emb <- Rtsne::Rtsne(gexpnorm)$Y
### k-means clustering 
# -- cluster using t-SNE embedding 
tot.witness = c()
for (i in 1:10){
  km <- kmeans(emb, centers=i)
  tot.witness = append(tot.witness, km$tot.withinss)
}
df.witness = data.frame(k=c(1:10), total.witness=tot.witness)
p00 = ggplot(df.witness, aes(x=k, y=tot.witness)) + geom_line() + 
  geom_point() + theme_minimal() + scale_x_discrete(limits=as.factor(c(1:10)))

km <- kmeans(emb, centers=6)
# -- plot k-means clusters in t-SNE space
df1 <- data.frame(emb, kmeans=as.factor(km$cluster))
p1 = ggplot(df1) + geom_point(aes(x=X1, y=X2, col=kmeans), 
                              size=0.1, alpha=0.5) + theme_minimal()
p1

# -- plot k-means clusters in physical space
df <- data.frame(data, kmeans=as.factor(km$cluster))
head(df)
p0 = ggplot(df) + geom_point(aes(x=x, y=y, col=kmeans), 
                             size=0.1, alpha=0.5) + theme_minimal()
p0

p00 + p1 + p0

### differential expression analysis
# -- Wilcoxon test and p-values 
# -- cluster 1
cluster_of_interest = 1
kmcluster = as.factor(km$cluster)

pvals <- sapply(colnames(gexp), function(p) {
  print(p)
  test <- wilcox.test(gexpnorm[kmcluster == cluster_of_interest, p], 
                 gexpnorm[kmcluster != cluster_of_interest, p],)
  test$p.value
})

fc <- sapply(colnames(gexp), function(p) {
  print(p)
  mean(gexpnorm[kmcluster == cluster_of_interest, p])/mean(gexpnorm[kmcluster != cluster_of_interest, p])
})

df2 <- data.frame(pv = -log10(pvals), log2fc = log2(fc), label=colnames(gexp))
head(df2)
p3 = ggplot(df2) + geom_point(aes(x = log2fc, y = pv), size=1) + 
  geom_text(aes(x = log2fc, y = pv, label=label), size=2.5, 
            hjust=-0.1, vjust=1.5, angle=300) +
  xlim(-2.5, 0.7) + ylim(0, 400)
df <- data.frame(emb, kmeans=as.factor(km$cluster == cluster_of_interest))
p4 = ggplot(df) + geom_point(aes(x=X1, y=X2, col=kmeans), 
                             size=0.1, alpha=0.7) + theme_minimal() +
  scale_color_manual(breaks = c('FALSE', 'TRUE'), values=c("gray", "blue"),
                     labels = c( "others","selected"))

p5 = ggplot(data.frame(emb, gene = gexpnorm[,"Lyve1"])) + 
  geom_point(aes(x=X1, y=X2, col=gene), size=0.1, alpha=0.3) + 
  scale_color_continuous(type="viridis", direction=-1)

df <- data.frame(data[,1:2], kmeans=as.factor(km$cluster == cluster_of_interest))
p9 = ggplot(df) + geom_point(aes(x = x, y=y, col=kmeans), 
                             size=0.1, alpha=0.7) + theme_minimal() +
  scale_color_manual(breaks = c('FALSE', 'TRUE'), values=c("gray", "magenta"),
                     labels = c( "others","cluster of interest"))
(p3 + p4)/(p5 +p9)

# -- cluster 6
cluster_of_interest = 6
kmcluster = as.factor(km$cluster)

pvals <- sapply(colnames(gexp), function(p) {
  print(p)
  test <- wilcox.test(gexpnorm[kmcluster == cluster_of_interest, p], 
                      gexpnorm[kmcluster != cluster_of_interest, p],)
  test$p.value
})

fc <- sapply(colnames(gexp), function(p) {
  print(p)
  mean(gexpnorm[kmcluster == cluster_of_interest, p])/mean(gexpnorm[kmcluster != cluster_of_interest, p])
})

df2 <- data.frame(pv = -log10(pvals), log2fc = log2(fc), label=colnames(gexp))
head(df2)
p6 = ggplot(df2) + geom_point(aes(x = log2fc, y = pv), size=1) + 
  geom_text(aes(x = log2fc, y = pv, label=label), size=2.5, hjust=-0.1, vjust=1.5,
            angle=300) +
  xlim(-2.5, 0.7) + ylim(0, 400)

df <- data.frame(emb, kmeans=as.factor(km$cluster == cluster_of_interest))
p7 = ggplot(df) + geom_point(aes(x=X1, y=X2, col=kmeans), 
                             size=1, alpha=0.7) + theme_minimal() +
  scale_color_manual(breaks = c('FALSE', 'TRUE'), values=c("gray", "blue"),
                     labels = c( "others","selected"))

p8 = ggplot(data.frame(emb, gene = gexpnorm[,'CD45'])) + 
  geom_point(aes(x=X1, y=X2, col=gene), size=0.1, alpha=0.3) + 
  scale_color_continuous(type="viridis", direction=-1)

df <- data.frame(data[,1:2], kmeans=as.factor(km$cluster == cluster_of_interest))
p10 = ggplot(df) + geom_point(aes(x = x, y=y, col=kmeans), 
                             size=0.3, alpha=0.7) + theme_minimal() +
  scale_color_manual(breaks = c('FALSE', 'TRUE'), values=c("gray", "magenta"),
                     labels = c( "others","cluster of interest"))

(p00+p1)/(p3 | p4 | p5 | p9)/(p6|p7|p8|p10)

``` 

## external resources
[1] https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2019.00036/full
[2] https://www.ncbi.nlm.nih.gov/gene/10894
[3] https://www.sciencedirect.com/science/article/pii/S0301472X2030182X 
[4] https://v15.proteinatlas.org/ENSG00000198851-CD3E/tissue/spleen